pkg/csi/cinder/controllerserver.go

```diff
@@ -134,12 +135,28 @@ func (cs *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVol
 		}
 	}
 
+	// Volume Encryption
+	encrypted := (req.GetParameters()["encrypted"] == "true")
+
+	if encrypted && volType == "" {
+		volType = cinderCSIDefaultEncryptedVolumeType
+	}
+
 	vol, err := cloud.CreateVolume(volName, volSizeGB, volType, volAvailability, snapshotID, sourcevolID, &properties)
 
 	if err != nil {
 		klog.Errorf("Failed to CreateVolume: %v", err)
 		return nil, status.Errorf(codes.Internal, "CreateVolume failed with error %v", err)
+	} else if encrypted && !vol.Encrypted {
+		klog.Errorf("Failed to CreateVolume with encryption enabled. Check volume type set.")
+
+		err := cloud.DeleteVolume(vol.ID)
+		if err != nil && !cpoerrors.IsNotFound(err) {
+			klog.Errorf("Failed to CreateVolume (cleanup): %v", err)
+			return nil, status.Errorf(codes.Internal, "CreateVolume with encryption enabled cleanup failed with error %v", err)
+		}
 
+		return nil, status.Errorf(codes.InvalidArgument, "Failed to CreateVolume with encryption enabled. Check volume type set.")
```

Comment:
Even if we agree that this is a good idea, we should check the volume type up front, not "try" creating a volume of that volume type to check if it's encrypted.